simulation/amaru-sim/src/simulator/simulate.rs

```diff
@@ -262,11 +291,105 @@ mod tests {
         assert_eq!(world.run_world(), &Vec::new());
     }
 
-    // NOTE: This will be replaced with a test that uses an in-process NodeHandle later, which
-    // avoids the problem of locating and spawning the binary.
+    #[test]
+    #[should_panic]
+    fn simulate_pure_stage_echo() {
+        let config = Default::default();
+
+        let number_of_nodes = 1;
+
+        let spawn: fn() -> NodeHandle = || {
+            println!("*** Spawning node!");
+            let mut network = SimulationBuilder::default();
+            let stage = network.stage(
+                "echo",
+                async |(mut state, out), msg: Envelope<EchoMessage>, eff| {
+                    if let EchoMessage::Echo { msg_id, echo } = &msg.body {
+                        state += 1;
+                        // Insert a bug every 5 messages.
+                        let reply;
+                        if state % 5 == 0 {
+                            reply = Envelope {
+                                src: msg.dest,
+                                dest: msg.src,
+                                body: EchoMessage::EchoOk {
+                                    msg_id: state,
+                                    in_reply_to: *msg_id,
+                                    echo: echo.to_string().to_uppercase(),
```

Comment:
I’d compute `echo` before constructing the Envelope, limiting the scope of the `if` (less code to check for differences)